includes/class-wc-payments-upe-checkout.php

```diff
@@ -161,6 +163,31 @@ public function get_payment_fields_js_config() {
 		return $payment_fields;
 	}
 
+	/**
+	 * This method extracts payment intent data and assumes failed state of operating flow, when the intent already
+	 * bound to an order or intent is processed but order state is not updated yet. Hence, we'll do some checks to
+	 * ensure buyers session stores a usable intent.
+	 *
+	 * Note: I know we should not enter a failed state, but real-live environments behaviour is unpredictable.
+	 *
+	 * @return string|null
+	 */
+	private function get_payment_intent_data_from_session() {
+		$intent_data = $this->gateway->get_payment_intent_data_from_session();
+		if ( $intent_data ) {
+			// Extract the intent and ensure it's not bound to an order yet nor being processed.
+			list( $cart_hash, $intent_id, $client_secret ) = explode( '-', $intent_data, 3 );
+			// TODO: should we inject this via constructor?
+			$order = ( new WC_Payments_DB() )->order_from_intent_id( $intent_id );
+			if ( ! $order && ! $this->gateway->is_intent_being_processed( $intent_id ) ) {
+				return $intent_data;
+			}
+			// Clean the cache, to recover it from failed state of our flows.
+			$this->gateway->remove_upe_setup_intent_from_session();
+		}
+		return null;
```

Comment:
```suggestion
			if ( $order || $this->gateway->is_intent_being_processed( $intent_id ) ) {
				// Clean the cache, to recover it from failed state of our flows.
				$this->gateway->remove_upe_setup_intent_from_session();
				$intent_data = null;
			}
		}
		return $intent_data;
```

Less complexity in the `if` statement and just one exit point. What do you think?

Reply:
Like the proposal, addressed in f05961ddc8beffc1ead89ebd05864d981c659b1d.